rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+runtime dimensions feature.
+
+The runtime dimensions feature is not a silver bullet, this is why the RFC will
+consider another option to reduce the primitive creation overhead.
+
+## Proposal
+
+There are two options to reduce primitive creation overhead:
+1. Extend the runtime dimensions feature for the relevant primitives
+2. Introduce a cache for kernels
+
+### Option 1: Cache for Kernels (recommended)
+
+The idea is to introduce a cache for kernels (hereinafter *cache*). The cache will
+be global similar to the primitive cache.
+
+Pros:
+* Doesn't require users to modify their code to make the feature work
+* Eliminates the kernel creation overhead completely when it's possible
+* Doesn't require to expose kernel implementation details. For example, for
+the runtime dimensions feature we will have to document what dimensions can
+be runtime without introducing performance regressions
+
+Cons:
+* Primitive and primitive descriptor creation overhead will stay
+
+#### Key and Value
+
+The cache implementation will reside in the `common` part of the library. In order
+to abstract away `key` and `value` entities we will need to introduce the
+following abstractions:
+
+```cpp
+namespace kernel_cache {
+// Key
+struct key_t {
+    key_t(key_impl_t *impl);
+    virtual ~key_t() = default;
+
+    bool operator==(const key_t &other) const;
+    size_t hash() const;
+private:
+    std::shared_ptr<key_impl_t> impl_;
+};
+
+// The `key_impl_t` class will be the base class for all `key` entities.
+// Each engine kind and vendor will be able to define it the way the need.
+struct key_impl_t {
+    key_impl_t() = default;
+    virtual ~key_impl_t() = default;
+
+    key_impl_t(const key_impl_t &) = delete;
+    key_impl_t &operator=(const key_impl_t &) = delete;
+
+    virtual bool compare(const key_impl_t *key_impl) const = 0;
+    virtual size_t hash() const = 0;
+};
+
+// The `value_t` class will be the base class for all `value` entities.
+// Each engine kind and vendor will be able to define it the way the need.
+struct value_t {
+    value_t(value_impl_t *impl);
+    virtual ~value_t() = default;
+    std::shared_ptr<value_impl_t> &impl() const;
+private:
+    std::shared_ptr<value_impl_t> impl_;
+};
+
+struct value_impl_t {
+    value_impl_t() = default;
+    virtual ~value_impl_t() = default;
+
+    value_impl_t(const value_impl_t &) = delete;
+    value_impl_t &operator=(const value_impl_t &) = delete;
+};
+} // kernel_cache
+```
+
+#### Example
+
+Below is a pseudo-code that demonstrates the idea of using the aforementioned
+abstractions for GPU engine kind.
+```cpp
+// GPU specific `key` and `value` implementations.
+struct gpu_key_t : public key_impl_t {
+    // This could be a serialized configuration structure or anything else.
+    std::vector <uint8_t> data;
+
+    compare(const key_impl_t *key_impl) const override {
+        const auto *typed_key = static_cast<const gpu_key_impl_t *>(key_impl);
+        return data == typed_key->data;
```

Comment:
With kernel cache, we will introduce some overhead in kernel seek.
And, primitive cache will not 1:1 mapping with kernel cache.
The penalty for latency in single batch workloads will be introduced by primitive cache seek and plus kernel cache seek?

Reply:
I guess the "kernel seek" would only happen once and saved in the primitive in later runs?

Reply:
The kernel cache lookup operation is performed only once during a primitive creation to fetch a kernel and store it in the primitive.
* If a primitive is found in the primitive cache then kernel cache won't be used
* If a primitive is not found in the primitive cache then we will have to create it as usually but the kernel generation overhead will be eliminated or more precisely replaced by the kernel cache lookup operation overhead